### PR TITLE
Find Unlinked files should ignore Thumbs.db, etc

### PR DESCRIPTION
**Pull request by prashant1712**
Fixes https://github.com/koppor/jabref/issues/373.

Reads the .gitignore file in the root directory of the library path and obeys it according to the option selected in the UI dialog box.

- [x] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [x] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [x] Screenshots added in PR description (for UI changes)
- [x] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [ ] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.
Screenshot 2022-05-16 at 5 09 16 AM

### CODE REVIEW

File: src/main/java/org/jabref/gui/externalfiles/FileFilterUtils.java [hunk 1]
```diff
@@ -75,7 +76,28 @@ public static boolean filterByDate(Path path, DateRange filter) {
...
     }
 
-    /* Sorts a list of Path objects according to the last edited date 
+    /* Returns true if a file should be ignored by .gitignore */
+    public static boolean filterForUnlinkedFiles(Path path, FileIgnoreUnlinkedFiles unlinkedFileFilter) {
+        if (unlinkedFileFilter.equals(FileIgnoreUnlinkedFiles.DEFAULT)) {
+            try {
+                FILEIGNORESET = FileIgnoreUnlinkedFiles.getIgnoreFileSet();
+                String fileExtension = "";
+                int i = path.toString().lastIndexOf('.');
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/gui/externalfiles/FileFilterUtils.java:85]: You can reuse FileUtil.getFileExtension() here

File: CHANGELOG.md [hunk 2]
```diff
@@ -13,6 +13,7 @@ Note that this project **does not** adhere to [Semantic Versioning](http://semve
 
 - We added a fetcher for [Biodiversity Heritage Library)](https://www.biodiversitylibrary.org/) [8539](https://github.com/JabRef/jabref/issues/8539)
 - We added support for multiple messages in the snackbar. [#7340](https://github.com/JabRef/jabref/issues/7340)
+- We added an extra option in the 'Find Unlinked Files' dialog view to ignore unnecessary files like Thumbs.db, DS_Store, etc. [#373](https://github.com/koppor/jabref/issues/373)
```
Comment (Siedlerchr) [inline CHANGELOG.md:16]:
```suggestion
- We added an extra option in the 'Find Unlinked Files' dialog view to ignore unnecessary files like Thumbs.db, DS_Store, etc. [koppor#373](https://github.com/koppor/jabref/issues/373)
```

**Comment by prashant1712** [comment]
Thank you for your suggestions. I will make the changes and commit it soon.

File: .gitignore [hunk 3]
```diff
@@ -239,7 +239,7 @@ fabric.properties
 *.zip
 *.tar.gz
 *.rar
-
+AAGenerator.cpp
```
Comment (calixtus) [inline .gitignore:242]: this seems to be an artifact of another programming project?

File: .gitignore [hunk 4]
```diff
@@ -418,7 +418,7 @@ Thumbs.db
 Thumbs.db:encryptable
 ehthumbs.db
 ehthumbs_vista.db
-
+data.tsv
```
Comment (calixtus) [inline .gitignore:421]: also an artifact?

File: src/main/java/org/jabref/gui/externalfiles/FileIgnoreUnlinkedFiles.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    private final String fileIgnoreOption;
+    FileIgnoreUnlinkedFiles(String fileIgnoreOption) {
+        this.fileIgnoreOption = fileIgnoreOption;
+    }
+
+    public String getFileIgnoreOption() {
+        return fileIgnoreOption;
+    }
+
+    public static Set<String> getIgnoreFileSet() {
```
Comment (calixtus) [inline src/main/java/org/jabref/gui/externalfiles/FileIgnoreUnlinkedFiles.java:33]: I'm really not a fan of these static methods. Keep the scope small and only use statics or globals if really neccessary. I think this method rather belongs in the UnlinkedFilesDialogViewModel

**Review by calixtus — CHANGES_REQUESTED** [review]
Great work. Some minor remarks, but looks already very good!

File: src/main/java/org/jabref/gui/externalfiles/FileFilterUtils.java [hunk 6]
```diff
@@ -8,15 +8,19 @@
...
+import org.jabref.logic.util.io.FileUtil;
+
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 public class FileFilterUtils {
 
     private static final Logger LOGGER = LoggerFactory.getLogger(FileFilterUtils.class);
-    
+    private static Set<String> FILEIGNORESET = null;
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/gui/externalfiles/FileFilterUtils.java:22]: I would initliaize this e.g = new HashSet<>() to prevent potentital NPEs

**Comment by Siedlerchr** [comment]
I did some refactorings to improve the code:
1. I moved down the reading of the ignore file from the dialog view model to the "logic" as it does not really depend on the dialog.
2. A general thing: Try to pass dependencies always down through the constructor (Inversion of Control). This makes it easier to test or mock the class and provides better encapsulation
3. I added a check to make sure the gitignore File is there before reading it. Relying on exceptions for control flows is a bad design idea. As the name implies, exceptions should be there when something unusual happens.


Now I would like you to add a test for new method FileFilterUtils and maybe as well for the UnlinkedFilesCrawler. That would be cool

**Comment by prashant1712** [comment]
@Siedlerchr Thank you for your help. Really appreciate the refactoring to improve the code, helped us in understanding the code better. One small thing, you changed the code to read the .gitignore file from the system of the user which i don't think is appropriate in our case. I'm changing it back to read it from the source directory of Jabref. Do let us know your opinion on this.

**Comment by Siedlerchr** [comment]
Yeah, that's one thing I wasn't sure about. I thought it makes sense to check the folder you selected to search for the files and the subfolders if there is any gitignore file inside and use this then?

**Comment by prashant1712** [comment]
Yeah, i'm not really sure what to do now. Which scenario are we going to take: Searching for the .gitignore file in the directory selected by the user or picking up the .gitignore file from the source directory of JabRef. We initially thought that it makes sense to ignore the extensions and file names mentioned in the .gitignore file of JabRef as it gives a consistent behaviour across all users but I'm quite confused right now and would really appreciate a solution to this.

**Comment by Siedlerchr** [comment]
Let's ask @koppor  as it was his issue...

File: src/main/java/org/jabref/gui/externalfiles/UnlinkedFilesCrawler.java [hunk 7]
```diff
@@ -1,13 +1,17 @@
 package org.jabref.gui.externalfiles;
 
+import java.io.BufferedReader;
 import java.io.File;
 import java.io.IOException;
 import java.nio.file.DirectoryStream.Filter;
 import java.nio.file.Files;
 import java.nio.file.Path;
```
Comment (calixtus) [inline src/main/java/org/jabref/gui/externalfiles/UnlinkedFilesCrawler.java:8]: I just mentioned you are mixing pre-java7-io api and nio. Is there a special reason?

**Comment by koppor** [comment]
> Yeah, i'm not really sure what to do now. Which scenario are we going to take: Searching for the .gitignore file in the directory selected by the user

Yes, do this.

**In addition** following files / directories are always ignored:

- `.git`
- `Thumbs.db`

This list is hard-coded in Java. No additional text file. No option for configuration.

**Comment by prashant1712** [comment]
Understood. Thank you @koppor for the clarification. I'll make the changes accordingly.

**Comment by Siedlerchr** [comment]
Please also include .DS_Store (A kind of desktop.ini equivalent for mac) in the default ignored files

**Comment by calixtus** [comment]
Any update here?

**Comment by prashant1712** [comment]
Hi @calixtus, i am recovering from a medical condition right now. Will complete the changes when i'm recovered. Thank you for understanding.

**Comment by calixtus** [comment]
Of course, get well soon.

**Comment by koppor** [comment]
@prashant1712 In each dev call, we choose a PR (more or less randomly) to finish it. Today, we worked 2 hours on this one. the intermediate result is there. However, the tests still fail. We need to continue here.

**Pull request merged**